src/librustc_typeck/variance/solve.rs

```diff
@@ -85,12 +85,19 @@ impl<'a, 'tcx> SolveContext<'a, 'tcx> {
         self.terms_cx.inferred_starts.iter().map(|(&id, &InferredIndex(start))| {
             let def_id = tcx.hir().local_def_id_from_hir_id(id);
             let generics = tcx.generics_of(def_id);
+            let count = generics.count();
 
-            let mut variances = solutions[start..start+generics.count()].to_vec();
-
+            let mut variances = solutions[start..(start + count)].to_vec();
             debug!("id={} variances={:?}", id, variances);
 
-            // Functions can have unused type parameters: make those invariant.
+            // Const parameters are always invariant.
+            for (idx, param) in generics.params.iter().enumerate() {
+                if let ty::GenericParamDefKind::Const = param.kind {
+                    variances[idx] = ty::Invariant;
```

Comment:
Shouldn't this be `variances[generics.parent_count + idx] = ty::Invariant`? Since `count` also include parent generics while `params` do not.

Reply:
Not sure if you'd also want to change the variance of parent generics here. Maybe this code should be elsewhere?

Reply:
cc @nikomatsakis

Reply:
Ah, you're absolutely right. Thanks for spotting that!